### PR TITLE
APCu driver

### PR DESCRIPTION
**Pull request by neo22s**
Added APCu native driver. In PHP7 I was facing issues since seems apc_functions are not working as a droppin. I have created the APCu as an alternative if someone else have this problem.

### CODE REVIEW

**Comment by rjd22** [comment]
@neo22s could you explain what the difference is between the APC and APCu driver? And could you also update the documentation in `guide/cache`.

**Comment by neo22s** [comment]
apc_ functions now are apcu_ functions. 

Ok I will update the doc and add the new driver.

**Comment by neo22s** [comment]
http://php.net/manual/en/ref.apc.php
http://php.net/manual/en/ref.apcu.php

**Comment by rjd22** [comment]
Ah cool, to clear things up for the others:
- APC is opcode cache and data store
- APCu is only data store
- OPcache is only opcode cache

**Comment by neo22s** [comment]
Damn I committed the guide changes directly on kohana/3.3/develop sorry https://github.com/kohana/cache/commits/3.3/develop

I have such a mess, I am updating 4 repos at same time that are using this, sorry.

**Comment by rjd22** [comment]
@neo22s don't mind they are only guide changes :)

File: classes/Kohana/Cache/Apcu.php [hunk 1]
```diff
@@ -0,0 +1,166 @@
+<?php defined('SYSPATH') or die('No direct script access.');
+/**
+ * [Kohana Cache](api/Kohana_Cache) APCu driver. Provides an opcode based
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: APCu is not a opcode based driver

**Comment by neo22s** [comment]
I know but aaaaaaargg bothers me!

I think we are safe to merge it I have tested the drivers in local and staging on our company.

File: classes/Kohana/Cache/Apcu.php [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+ * @package    Kohana/Cache
+ * @category   Base
+ * @author     Kohana Team
+ * @copyright  (c) 2009-2012 Kohana Team
+ * @license    http://kohanaphp.com/license
+ */
+class Kohana_Cache_Apcu extends Cache implements Cache_Arithmetic {
+
+	/**
+	 * Check for existence of the APC extension This method cannot be invoked externally. The driver must
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: checking for the `APCu` extension :)

File: classes/Kohana/Cache/Apcu.php [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+ * [Kohana Cache](api/Kohana_Cache) APCu driver. Provides an opcode based
+ * driver for the Kohana Cache library.
+ *
+ * ### Configuration example
+ *
+ * Below is an example of an _apcu_ server configuration.
+ *
+ *     return array(
+ *          'apcu' => array(                          // Driver group
+ *                  'driver'         => 'apcu',         // using APC driver
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: // using APCu driver

File: classes/Kohana/Cache/Apcu.php [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+ *
+ * Name           | Required | Description
+ * -------------- | -------- | ---------------------------------------------------------------
+ * driver         | __YES__  | (_string_) The driver type to use
+ *
+ * ### System requirements
+ *
+ * *  Kohana 3.0.x
+ * *  PHP 5.2.4 or greater
+ * *  APC PHP extension
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: ACPu PHP extension

File: classes/Kohana/Cache/Apcu.php [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+			throw new Cache_Exception('PHP APCu extension is not available.');
+		}
+
+		parent::__construct($config);
+	}
+
+	/**
+	 * Retrieve a cached value entry by id.
+	 *
+	 *     // Retrieve cache entry from apc group
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: apcu group

File: classes/Kohana/Cache/Apcu.php [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+	 *     $data = Cache::instance('apcu')->get('foo');
+	 *
+	 *     // Retrieve cache entry from apc group and return 'bar' if miss
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: apcu group

File: classes/Kohana/Cache/Apcu.php [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+		return $success ? $data : $default;
+	}
+
+	/**
+	 * Set a value to cache with id and lifetime
+	 *
+	 *     $data = 'bar';
+	 *
+	 *     // Set 'bar' to 'foo' in apc group, using default expiry
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: apcu group

File: classes/Kohana/Cache/Apcu.php [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+	 *     Cache::instance('apcu')->set('foo', $data);
+	 *
+	 *     // Set 'bar' to 'foo' in apc group for 30 seconds
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: apcu group

File: classes/Kohana/Cache/Apcu.php [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+			$lifetime = Arr::get($this->_config, 'default_expire', Cache::DEFAULT_EXPIRE);
+		}
+
+		return apcu_store($this->_sanitize_id($id), $data, $lifetime);
+	}
+
+	/**
+	 * Delete a cache entry based on id
+	 *
+	 *     // Delete 'foo' entry from the apc group
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: apcu group

File: classes/Kohana/Cache/Apcu.php [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	}
+
+	/**
+	 * Delete all cache entries.
+	 *
+	 * Beware of using this method when
+	 * using shared memory cache systems, as it will wipe every
+	 * entry within the system for all clients.
+	 *
+	 *     // Delete all cache entries in the apc group
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: apcu group

File: classes/Kohana/Cache/Apcu.php [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+	 * @param   int       step value to decrement by
+	 * @return  integer
+	 * @return  boolean
+	 */
+	public function decrement($id, $step = 1)
+	{
+		return apcu_dec($id, $step);
+	}
+
+} // End Kohana_Cache_Apc
```
Comment (rjd22) [inline classes/Kohana/Cache/Apcu.php]: // End Kohana_Cache_Apcu

File: classes/Kohana/Cache/Apcu.php (hunk 2 above)
Comment (neo22s) [inline classes/Kohana/Cache/Apcu.php]: Yeah love APC as cache for variable=>value, super fast and easy.

**Comment by rjd22** [comment]
@neo22s a few nitpicks there where instances of apc are still there that need to be replaced to apcu. Could you change this?

After this you get a :+1: from me

**Comment by neo22s** [comment]
@rjd22 dont follow, what is missing? ok I tihnk I understand some words in the comments APC for APCu sure no problem.

**Comment by neo22s** [comment]
updated found an error on the functions, I have reviewed the docs for all the other methods. was the only one not working apcu_clear_cache

**Comment by acoulton** [comment]
@neo22s thanks for this, can you also add tests before it's merged? Hopefully you can just copy tests/cache/arithmetic/ApcTest.php[https://github.com/kohana/cache/blob/3.3/develop/tests/cache/arithmetic/ApcTest.php] and tweak it.

**Comment by neo22s** [comment]
done. damn is so complex to collaborate! thanks

**Comment by enov** [comment]
> damn is so complex to collaborate!

Indeed :smile:

**Comment by neo22s** [comment]
I'm sorry to be so faulty but the lack of time I have.... having crazy times I have a newborn baby, company is having huge peak of work and not sleeping much, so I can not invest more time. regards.

**Comment by enov** [comment]
> I have a newborn baby

Ah :baby: congrats!

**Comment by neo22s** [comment]
Thanks! 5 months, but damn so much work! ;)

Saludos / Regards, Chema
GarridoDiaz.com http://garridodiaz.com

On Thu, Dec 17, 2015 at 9:49 AM, Samuel Demirdjian <notifications@github.com

**Comment by rjd22** [comment]
@neo22s @enov what needs to happen to get this merged? It's a pity to leave these almost finished pull unmerged.

**Comment by neo22s** [comment]
@rjd22 right now nothing. If I remember well. I can also do the merge and finish it I have permissions, but if you want to take a look fee free ;)

**Comment by enov** [comment]
@rjd22 @neo22s it probably needs squashing into a single commit, but I don't mind. @acoulton?

**Comment by acoulton** [comment]
I think it was just waiting on the two comments corrections proposed by @rjd22 that are still showing in the discussion above - otherwise it looks good to me. @enov can you do them and then this can be merged?

**Comment by neo22s** [comment]
Looks good guys move forward! we look like a big corporation that never moves fast enough ;) it's a pity ;)

**Comment by acoulton** [comment]
@neo22s well we're all volunteers, and it's been Christmas holidays... and @rjd22 flagged the issues that are outstanding about 15 minutes after you submitted the pull request.... :grinning:

**Comment by neo22s** [comment]
@acoulton I know! and its amazing what you have done ;) its just that we all ask permission to everyone and get complicated heheheh kudos guys ;)

**Comment by enov** [comment]
@acoulton the two outstanding comments corrections proposed by @rjd22 are fixed now.

**Comment by acoulton** [comment]
Thanks @enov @rjd22 @neo22s

**Comment by neo22s** [comment]
awesome! thanks

**Pull request merged**